File: 프로그래머스/lv3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<bool> visited;
int num;

void dfs(vector<vector<int>> &board, int cur) {
    visited[cur] = true;
    for(int i = 0; i < num; i++) {
        if(board[cur][i] && !visited[i]) {
            dfs(board, i);
        }
    }
    return;
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    num = n;
    visited = vector<bool> (n, false);
    
    for(int i = 0; i < n; i++) {
        if(!visited[i]) {
            answer++;
            dfs(computers, i);
        }
    }
    
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
def solution(n, computers):
    answer = 0
    
    adjs = [[] for _ in range(n)]
    is_visited = [False] * n
    
    def DFS(idx):
        for nxt in adjs[idx]:
            if not is_visited[nxt]:
                is_visited[nxt] = True
                DFS(nxt)
    
    for i, rows in enumerate(computers):
        for j, connected in enumerate(rows):
            if connected:
                adjs[i].append(j)
                
    for idx in range(n):
        if is_visited[idx]:
            continue
        
        is_visited[idx] = True
        DFS(idx)
        answer += 1
    
    return answer